src/state/opamSwitchState.ml

```diff
@@ -1316,26 +1318,30 @@ let dependencies_filter_to_formula_t ~build ~post st nv =
   in
   OpamFilter.filter_formula ~default:true env
 
-let dependencies_t base_deps_compute deps_compute
-    ~depopts ~installed ?(unavailable=false) universe packages =
+let dependencies_t st base_deps_compute deps_compute
+    ~depopts ~installed ?(unavailable=false) packages =
   if OpamPackage.Set.is_empty packages then OpamPackage.Set.empty else
   let base =
     packages ++
-    if installed then  universe.u_installed
-    else if unavailable then universe.u_packages
-    else universe.u_available
+    if installed then st.installed
+    else if unavailable then st.packages
+    else Lazy.force st.available_packages
   in
   log ~level:3 "dependencies packages=%a"
     (slog OpamPackage.Set.to_string) packages;
   let timer = OpamConsole.timer () in
   let base_depends =
     let filter = base_deps_compute base in
+    let get_deps = get_deps st ~force_dev_deps:false ~test:false ~doc:false ~dev_setup:false ~requested_allpkgs:packages in
+    let opams = OpamPackage.Set.fold (fun pkg opams -> OpamPackage.Map.add pkg (OpamPackage.Map.find pkg st.opams) opams) base OpamPackage.Map.empty in
```

Comment:
is there not a function to do that somewhere already? It feels possibly like a very common pattern in opam's code but I wasn't able to find an example